Update dis_macro and add ref dis injection

garethj2 opened the pull request:
Two changes here...

1. **dis_macro**: now returns a `Cow` for the `get_value` function passed into it. This gives more flexibility when using the API. Please note, no version has been published since this last change so it can't be considered a breaking change.
2. **ref dis injection**: I needed someone to inject `Ref` display name values into decoded objects. This is required in a haystack server when one is dynamically calculating a record's `display` value and needs to share it with all record's in the database in an efficient manner. I admit I'm not 100% happy with this. I couldn't find a nice efficient way to do some sort of configuration injection into a Serde deserializer. Regardless, I've isolated this to a specific part of the library so it doesn't pollute the core `Ref` value type. I think this really could be enhanced further by making the core API able to share references more easily and hence use proper interning for granular haystack data types.

garethj2 committed:
Update dis_macro and add ref dis injectiond

garethj2 committed:
Update comments

garethj2 committed:
Add to encoder

hecsalazarf commented on src/haystack/encoding/json/decode.rs:
```diff
@@ -103,7 +104,19 @@ impl<'de> Deserialize<'de> for Ref {
         match val {
-            HVal::Ref(val) => Ok(val),
+            HVal::Ref(val) => {
+                let dis = get_ref_dis_from_factory(val.value.as_str(), val.dis.as_deref())
```
It seems an owned string is needed if `dis != val.dis` only. This could prevent unnecessary allocations by taking advantage of `Cow`:
```rust
let dis = get_ref_dis_from_factory(val.value.as_str(), val.dis.as_deref());

Ok(if dis.as_deref() == val.dis.as_deref() {
    val
} else {
    Ref {
        value: val.value,
        dis: dis.map(|d| d.into_owned()),
    }
})
```

garethj2 replied:
Agreed. Great idea.

hecsalazarf commented on src/haystack/encoding/json/decode.rs:
```diff
@@ -377,16 +390,26 @@ fn parse_number(dict: &Dict) -> Result<HVal, JsonErr> {
 fn parse_ref(dict: &Dict) -> Result<HVal, JsonErr> {
```
`parse_ref` could be rewritten to minimize expressions repetition:
```rust
fn parse_ref(dict: &Dict) -> Result<HVal, JsonErr> {
    dict.get_str("val")
        .map(|val| {
            let dis_str = dict.get_str("dis").map(|d| d.as_str());
            let dis = get_ref_dis_from_factory(&val.value, dis_str).map(|val| val.into_owned());
            Ref {
                value: val.value.clone(),
                dis,
            }
            .into()
        })
        .ok_or_else(|| JsonErr::custom("Missing or invalid 'val'"))
}
````

hecsalazarf commented on src/haystack/encoding/json/encode.rs:
```diff
@@ -54,8 +57,12 @@ impl Serialize for Ref {
-            let dis = self.dis.clone();
+
+        let dis = get_ref_dis_from_factory(self.value.as_str(), self.dis.as_deref())
+            .map(|v| v.to_string())
```
nitpick: Prefer `v.into_owned()`. `to_string` is a method derived from the `Display` trait, meant to print types as formatted strings. The result is the same, but `Cow::into_owned` provides better semantics since the objective is to create an owned value, not a formatted string. Besides, `to_string` calls the formatting logic, whereas `into_owned` simply clones the wrapped value if required.

garethj2 commented:
Would appreciate your thoughts on this @rracariu . I just want to make sure you're happy with my ref display name injection addition.

garethj2 committed:
PR feedback

rracariu reviewed (commented):
I would go with a trait that implements the generic approach in the library, and then re-implement this trait in another crate using the new-type pattern, see [Using the Newtype Pattern to Implement External Traits on External Types](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types)
IMHO this would make this less fragile and most likely more performant as there are less indirections and lock access, in exchange of a little more coding effort in the consumer crate.

To summarize: crate a trait for this macro facility, implement it for `Value` and `Ref` in this library crate, use this default implementation for the ser/de, then in another crate that depends on this library, create wrappers for `Value` and `Ref` (List, Dict, Grid?) implement the dis macro trait for these wrappers and partially implement the Serde traits for serialize/deserialize parts where you do your specialization for whatever use case you have, and delegate the rest to the default `Value` and `Ref` types. Create wrappers for zinc if this is the case.

If you want, you could completely circumvent the library type crate and only use the wrappers in the consuming crate by using transparent references via the [Deref polymorphism](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html#deref-polymorphism)
I don't think this would be need it, as probably you only need this facility of ser/de parts, so in this case those wrapper structs would suffice.

rracariu commented on src/haystack/encoding/decode_ref_dis_factory.rs:
```diff
@@ -0,0 +1,65 @@
+type RefDisFactoryFunc<'a> = Box<dyn Fn(&str, Option<&str>) -> Option<Cow<'a, str>> + Send + Sync>;
+
+/// Holds the Ref factory function used for creating Refs.
+static REF_DIS_FACTORY_FUNC: RwLock<Option<RefDisFactoryFunc>> = RwLock::new(None);
```
I suspect that [AtomicPtr](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicPtr.html) would be a better choice here.

garethj2 replied:
Thanks for the tip. I think the `RwLock` looks a little cleaner though.

rracariu replied:
Yes, but it has more overhead.

garethj2 committed:
Remove ref display name injection + make decoder visitor public

garethj2 commented:
> I would go with a trait that implements the generic approach in the library, and then re-implement this trait in another crate using the new-type pattern, see [Using the Newtype Pattern to Implement External Traits on External Types](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types) IMHO this would make this less fragile and most likely more performant as there are less indirections and lock access, in exchange of a little more coding effort in the consumer crate.
> 
> To summarize: crate a trait for this macro facility, implement it for `Value` and `Ref` in this library crate, use this default implementation for the ser/de, then in another crate that depends on this library, create wrappers for `Value` and `Ref` (List, Dict, Grid?) implement the dis macro trait for these wrappers and partially implement the Serde traits for serialize/deserialize parts where you do your specialization for whatever use case you have, and delegate the rest to the default `Value` and `Ref` types. Create wrappers for zinc if this is the case.
> 
> If you want, you could completely circumvent the library type crate and only use the wrappers in the consuming crate by using transparent references via the [Deref polymorphism](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html#deref-polymorphism) I don't think this would be need it, as probably you only need this facility of ser/de parts, so in this case those wrapper structs would suffice.

Thanks for the great advice @rracariu . I've implemented this and it works very nicely! I've removed the ref display name injection from the code base. I did make the serde visitor decoder public so I could reuse when I create my own serde decoder visitor. I agree this is a far better approach for now. It keeps this library nice and clean.

rracariu reviewed (approved):
LGTM
